examples/pytorch/aml.py

```diff
@@ -0,0 +1,264 @@
+        scores = scores[mask]  # [batch_size]
+        pred = scores.argmax(dim=1)
+        loss = self.xent(scores, labels)
+        return loss, pred, labels
+
+
+def train_func(config: dict):
+    """Training loop for ray trainer."""
+    cora = Cora()
+
+    address = "localhost:9999"
+    _ = Server(address, cora.data_dir(), 0, 1)
+
+    def get_graph():
+        return DistributedClient([address])
+
+    config["get_graph"] = get_graph
+
+    train.torch.enable_reproducibility(seed=session.get_world_rank())
+
+    p = GATQueryParameter(
+        neighbor_edge_types=np.array([0], np.int32),
+        feature_idx=config["feature_idx"],
+        feature_dim=config["feature_dim"],
+        label_idx=config["label_idx"],
+        label_dim=config["label_dim"],
+    )
+    model = GAT(
+        in_dim=config["feature_dim"],
+        head_num=[8, 1],
+        hidden_dim=8,
+        num_classes=config["num_classes"],
+        ffd_drop=0.6,
+        attn_drop=0.6,
+        q_param=p,
+    )
+    model = train.torch.prepare_model(model)
+
+    optimizer = torch.optim.Adam(
+        filter(lambda p: p.requires_grad, model.parameters()),
+        lr=0.005 * session.get_world_size(),
+        weight_decay=0.0005,
+    )
+    optimizer = train.torch.prepare_optimizer(optimizer)
+
+    g = config["get_graph"]()
+    batch_size = 140
+    train_dataset = ray.data.read_text(f"{cora.data_dir()}/train.nodes")
+    train_dataset = train_dataset.repartition(train_dataset.count() // batch_size)
+    train_pipe = train_dataset.window(blocks_per_window=4).repeat(config["num_epochs"])
+    train_pipe = train_pipe.map_batches(
+        lambda idx: model.q.query_training(g, np.array(idx))
+    )
+
+    test_dataset = ray.data.read_text(f"{cora.data_dir()}/test.nodes")
+    test_dataset = test_dataset.repartition(1)
+    test_dataset = test_dataset.map_batches(
+        lambda idx: model.q.query_training(g, np.array(idx))
+    )
+    test_dataset_iter = test_dataset.repeat(config["num_epochs"]).iter_epochs()
+
+    for epoch_pipe in train_pipe.iter_epochs():
+        model.train()
+        losses = []
+        for batch in epoch_pipe.iter_torch_batches(batch_size=batch_size):
+            loss, score, label = model(batch)
+            optimizer.zero_grad()
+            loss.backward()
+            optimizer.step()
+
+            losses.append(loss.item())
+
+        model.eval()
+        batch = next(next(test_dataset_iter).iter_torch_batches(batch_size=1000))
+        loss, score, label = model(batch)
+        test_scores = [score]
+        test_labels = [label]
+
+        session.report(
+            {
+                model.metric_name(): model.compute_metric(
+                    test_scores, test_labels
+                ).item(),
+                "loss": np.mean(losses),
+            },
+        )
+
+
+if __name__ == "__main__":
+    from azureml.core import Workspace
+    from ray_on_aml.core import Ray_On_AML
+
+    try:
+        ws = Workspace.from_config()
```

Comment:
Synced offline: please make this example more useful, current version is a straightforward application of [the existing RayOnML example](https://github.com/microsoft/ray-on-aml/blob/master/examples/job/aml_job.py). However, most of the users will probably need to set up a separate graph engine instances and somehow connect workers to it.